### PR TITLE
Add `vmodl:verify` and `vmodl:generate` rake tasks

### PR DESCRIPTION
**Pull request by agrare**
TODO:
- [x] Move `devel/verify-vim-wsdl.rb` into a class

### CODE REVIEW

File: lib/tasks/vmodl_helper.rb [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        opt :vmodl, 'Path to the vmodl.db', type: :string, default: 'vmodl.db'
+        banner <<~EOS
+          Usage:
+          rake vmodl:generate -- --wsdl=path/to/wsdl
+        EOS
+      end
+
+      new(vmodl_path: opts[:vmodl], wsdl_path: opts[:wsdl]).generate!
+    end
+  end
```
Comment (Fryguy) [inline lib/tasks/vmodl_helper.rb:40]: There some redundancy here that I think can be cleaned up.
```suggestion
  class << self
    private def run!(mode, argv)
      opts = Optimist.options(argv) do
        educate_on_error
        opt :wsdl, 'Path to the vsphere-ws wsdl file', type: :string, required: true
        opt :vmodl, 'Path to the vmodl.db', type: :string, default: 'vmodl.db'
        banner <<~EOS
          Usage:
          rake vmodl:#{mode} -- --wsdl=path/to/wsdl
        EOS
      end
      new(vmodl_path: opts[:vmodl], wsdl_path: opts[:wsdl]).send("#{mode}!")
    end

    def verify!(argv)
      run!(:verify, argv)
    end

    def generate!(argv)
      run!(:generate, argv)
    end
  end
```

File: lib/tasks/vmodl.rake [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  task :generate do
+    VmodlHelper.generate!(argv)
+  end
+
+  private
+
+  # Pass everything after the `--` to Optimist
+  # E.g. rake vmodl:verify -- --wsdl=path/to/wsdl
+  def argv
+    (i = ARGV.index('--')) ? ARGV.slice((i + 1)..-1) : ARGV.dup
```
[the remark this answers is not in the document]
Comment (agrare) [reply lib/tasks/vmodl.rake:20]: Yeah true, originally i had the optimist arg parsing in here but moved it to the VmodlHelper

File: lib/tasks/vmodl.rake [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+require_relative './vmodl_helper'
+
+namespace :vmodl do
+  desc 'Verify vmodl.db'
+  task :verify do
+    VmodlHelper.verify!
+  end
+
+  task :generate do
```
[the remark this answers is not in the document]
Comment (agrare) [reply lib/tasks/vmodl.rake:12]: Thanks, added

File: lib/tasks/vmodl_helper.rb [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+    private
+
+    def run!(mode)
+      vmodl, wsdl = options(mode).values_at(:vmodl, :wsdl)
+      new(vmodl_path: vmodl, wsdl_path: wsdl).send("#{mode}!")
+    end
+
+    def options(rake_task)
+      argv = (i = ARGV.index('--')) ? ARGV.slice((i + 1)..-1) : ARGV.dup
```
[the remark this answers is not in the document]
Comment (agrare) [reply lib/tasks/vmodl_helper.rb:29]: I'll try that, pulled this from https://github.com/ManageIQ/manageiq/blob/master/lib/tasks/evm_rake_helper.rb#L38-L41

File: lib/tasks/vmodl_helper.rb [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    @vmodl = load_vmodl(@vmodl_path)
+    @wsdl  = load_wsdl(@wsdl_path)
+  end
+
+  def verify!
+    # Loop through the ComplexTypes in the WSDL and compare their types
+    # to the types which are defined in the vmodl.db
+    wsdl_types_by_name.each_value do |type|
+      type_name = type.name.name
+      next if type_name.match?(/^ArrayOf/) || type_name.match(/RequestType$/)
```
Comment (Fryguy) [inline lib/tasks/vmodl_helper.rb:55]:
```suggestion
      next if type_name.start_with?("ArrayOf") || type_name.end_with?("RequestType")
```

File: lib/tasks/vmodl_helper.rb [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+  def dump_vmodl!
+    File.write(@vmodl_path, Marshal.dump(@vmodl))
+  end
+
+  private
+
+  def wsdl_types_by_name
+    @wsdl_types_by_name ||= @wsdl.collect_complextypes
+      .reject   { |type| type.name.name.match?(/^ArrayOf|RequestType$/) }
```
Comment (Fryguy) [inline lib/tasks/vmodl_helper.rb:166]: Or use this same Regex up above in line 55

File: lib/tasks/vmodl_helper.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (Fryguy) [reply lib/tasks/vmodl_helper.rb:29]: Merged anyway - these can be follow ups because they are super minor.

File: lib/tasks/vmodl_helper.rb (hunk 5 above)
Comment (agrare) [reply lib/tasks/vmodl_helper.rb:55]: Actually this isn't needed anymore, being filtered out by https://github.com/ManageIQ/rbvmomi2/pull/9#discussion_r727459548

**Pull request merged**